### PR TITLE
Do Not Merge, Not yet :-) GSoC FEM Elmer project

### PR DESCRIPTION
**Pull request by berndhahnebach**
Replaces https://github.com/FreeCAD/FreeCAD/pull/945 since a rebase of this does not make sence at all anymore ...

Still it is not finished yet, but the implementation of the tools and the unit test needed to make a PR are done. Some testing is still needed.

forum: https://forum.freecadweb.org/viewtopic.php?f=18&t=22576
branch which has rewritten history (but is equal to this one in sence of diff) and will be used for PR eventually: https://github.com/berndhahnebach/FreeCAD_bhb/commits/bhbnewsolver

Bernd

### CODE REVIEW

**Comment by berndhahnebach** [comment]
the new elmer unit test fails on Python3 :-( I do not have compiled FreeCAD with Py3, seams it is time to do ... 

https://travis-ci.org/FreeCAD/FreeCAD/builds/299084331?utm_source=github_status&utm_medium=notification

**Comment by sasobadovinac** [comment]
`ERROR: test_static_freq_analysis (TestFem.FemCcxAnalysisTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Mod/Fem/TestFem.py", line 676, in test_static_freq_analysis
    ret = compare_files(test_file_dir_elmer + 'ELMERSOLVER_STARTINFO', static2_analysis_dir + 'ELMERSOLVER_STARTINFO')
  File "/usr/local/Mod/Fem/TestFem.py", line 1110, in compare_files
    file2 = open(file_name2, 'r')
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/FEM_unittests/FEM_static2/ELMERSOLVER_STARTINFO'`

File: src/Mod/Fem/TestFem.py [hunk 1]
```diff
@@ -1076,6 +1072,20 @@ def fcc_print(message):
     FreeCAD.Console.PrintMessage('{} \n'.format(message))
 
 
+def get_defmake_count():
+    '''
+    count the def make in module ObjectsFem
+    could also be donn in bash with
```
Comment (luzpaz) [inline src/Mod/Fem/TestFem.py]: donn->done

Comment (berndhahnebach) [reply src/Mod/Fem/TestFem.py]: donn :-) :-)

**Comment by berndhahnebach** [comment]
any of the Py3 guys an idea?

**Comment by sasobadovinac** [comment]
@looooo @yorikvanhavre

**Comment by looooo** [comment]
The file is not existing which should be loaded is not existing. The error is maybe at the place where the file is written...

**Comment by sasobadovinac** [comment]
@looooo thanks for looking in to it, I asked for your help because the issue seems to come up only with py3

**Comment by looooo** [comment]
thats true but it's difficult to track... this should be the place where the file is created:
https://github.com/berndhahnebach/FreeCAD_bhb/blob/femnewsolver2/src/Mod/Fem/femsolver/elmer/writer.py#L139L142

**Comment by looooo** [comment]
https://github.com/berndhahnebach/FreeCAD_bhb/blob/femnewsolver2/src/Mod/Fem/femsolver/elmer/writer.py#L75

iteritems is removed in python3!

But this still doesn't solve the test. Now the subprocess gets aborted silently somewhere else. It would be nice if someone who knows how to work with Fem would work with python3. I can help in setting up a py3 environment with conda. This maybe takes less time, than I spend on this really nasty problem.

**Comment by berndhahnebach** [comment]
Thanks guys. I could tackle down a few too. I used travis :-) https://github.com/FreeCAD/FreeCAD/pull/1076/commits/34833de1f05c0fa97c2276a2356e848f3af5842b

**Comment by looooo** [comment]
Thanks for that hard extra work!
it was already hard to find some problems with a local build. But you managed to get it done with travis...thats insane ;)
I still believe setting up a python3 environment would make sense. All these problems are maybe easier to find with the gui. The tests didn't give good tracebacks. Also the tests do not track all implemented features.

File: src/Mod/Fem/Gui/DlgSettingsFemElmer.ui [hunk 2]
```diff
@@ -18,149 +18,194 @@
...
+             </size>
+            </property>
+            <property name="baseSize">
+             <size>
+              <width>0</width>
+              <height>0</height>
+             </size>
+            </property>
+            <property name="toolTip">
+             <string>Leave blank to use default Elmer elmer binary file</string>
```
Comment (luzpaz) [inline src/Mod/Fem/Gui/DlgSettingsFemElmer.ui]: @berndhahnebach should `Elmer elmer` be duplicated here ?

**Comment by berndhahnebach** [comment]
https://github.com/FreeCAD/FreeCAD/pull/1130

**Pull request closed without merging**